### ISSUE TITLE
GREATEST function with DATE type returns DATETIME result

### ISSUE THREAD

**Issue by espresso98** [pingcap/tidb#30261]
### 1. Minimal reproduce step (Required)

```sql
SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303);
```

### 2. What did you expect to see? 

In MySQL8.0,
```sql
mysql> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303);
+-----------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) |
+-----------------------------------------------------------+
| 2005-05-05                                                |
+-----------------------------------------------------------+
```
In MySQL5.7,
```sql
mysql> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303);
+-----------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) |
+-----------------------------------------------------------+
| 20050505                                                  |
+-----------------------------------------------------------+
```

### 3. What did you see instead 
```sql
TiDB> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303);
+-----------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) |
+-----------------------------------------------------------+
| 2005-05-05 00:00:00                                       |
+-----------------------------------------------------------+
```

### 4. What is your TiDB version? 
```sql
tidb_version(): Release Version: v5.2.2
Edition: Community
Git Commit Hash: da1c21fd45a4ea5900ac16d2f4a248143f378d18
Git Branch: heads/refs/tags/v5.2.2
UTC Build Time: 2021-10-20 06:03:29
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

**Comment by mjonss** [comment]
Duplicate of https://github.com/pingcap/tidb/issues/30264 ?

**Comment by windtalker** [comment]
Should be fixed by #31037
